### PR TITLE
IGNITE-13919 SQL Calcite: Optimization for COUNT(column) if column is indexed

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/AggregatesIntegrationTest.java [hunk 1]
```diff
@@ -108,6 +108,26 @@ public void testCountWithBackupsAndCacheModes() {
         assertQuery("select count(*) from person").returns(7L).check();
     }
 
+    /** */
+    @Test
+    public void testCountIndexedField() {
+        createAndPopulateIndexedTable(1, CacheMode.PARTITIONED);
+        assertQuery("select count(distinct descVal) from person").returns(3L).check();
```
Comment (Vladsz83) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/AggregatesIntegrationTest.java:115]: Duplicate 'assertQuery("select count(distinct descVal) from person").returns(3L).check();'

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/IndexScan.java [hunk 2]
```diff
@@ -305,6 +311,34 @@ private synchronized void release() {
...
+        return new BPlusTree.TreeRowClosure<IndexRow, IndexRow>() {
+            /** {@inheritDoc} */
+            @Override public boolean apply(
+                BPlusTree<IndexRow, IndexRow> tree,
+                BPlusIO<IndexRow> io,
+                long pageAddr,
+                int idx
+            ) throws IgniteCheckedException {
+                if (keyType != null && io instanceof InlineIO) {
+                    Boolean keyIsNull = keyType.isNull(pageAddr, io.offset(idx), ((InlineIO)io).inlineSize());
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/IndexScan.java:331]: We take first keyType velue from 'inlineKeyTypes'. Is it possible that one key value is null when some other is not?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/CacheIndexImpl.java [hunk 3]
```diff
@@ -153,9 +158,43 @@ public Index queryIndex() {
...
+
+                rowFilter = new BPlusTree.TreeRowClosure<IndexRow, IndexRow>() {
+                    @Override public boolean apply(
+                        BPlusTree<IndexRow, IndexRow> tree,
+                        BPlusIO<IndexRow> io,
+                        long pageAddr,
+                        int idx
+                    ) throws IgniteCheckedException {
+                        // If we have NULLS-FIRST collation, all values after first not-null value will be not-null,
+                        // don't need to check it with notNullRowFilter.
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/CacheIndexImpl.java:179]: Is there a specific test for this case with 'COUNT()'?

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/AggregatesIntegrationTest.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/AggregatesIntegrationTest.java:115]: FIxed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexCountRule.java [hunk 4]
```diff
@@ -54,18 +57,50 @@ private IndexCountRule(IndexCountRule.Config cfg) {
...
+            notNull = true;
+            int fieldIdx = argList.get(0);
+
+            if (!scan.requiredColumns().isEmpty())
+                fieldIdx = scan.requiredColumns().nth(fieldIdx);
+
+            for (IgniteIndex idx0 : table.indexes().values()) {
+                List<RelFieldCollation> fieldCollations = idx0.collation().getFieldCollations();
+
+                if (!fieldCollations.isEmpty() && fieldCollations.get(0).getFieldIndex() == fieldIdx) {
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexCountRule.java:94]: Currently, no. But perhaps in future we can add hash indexes without collation (we have now hash index for _KEY, but we don't use it).

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/CacheIndexImpl.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/CacheIndexImpl.java:179]: Actually it tested in testCountIndexedField (some fields have NULL value), but I've added yet another test with NULL and two indexed columns.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/types/NullableInlineIndexKeyType.java [hunk 5]
```diff
@@ -167,9 +174,9 @@ protected NullableInlineIndexKeyType(IndexKeyType type, short keySize) {
      * @param pageAddr Page address.
      * @param off Offset.
      *
-     * @return Inline value or {@code null} if value can't be restored.
+     * @return Inline value.
      */
-    protected abstract @Nullable T get0(long pageAddr, int off);
+    protected abstract T get0(long pageAddr, int off);
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/types/NullableInlineIndexKeyType.java:179]: There is still a code, checking the null: `IndexQueryCriteriaClosure#rowIsOutOfRange()`. Should it be fixed? There is no null returns any more?

Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/types/NullableInlineIndexKeyType.java:179]: `rowIsOutOfRange` checks `get` method. `get` can be null, but `get0` can't.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexCountRule.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexCountRule.java:94]: The rule's javadoc says:

/** Tries to optimize 'COUNT(*)' to use number of index records. */

Might be changed to smth like

/** Tries to optimize 'COUNT()' to use number of index records. */

Non only all (*)

[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexCountRule.java:94]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexKeyType.java [hunk 6]
```diff
@@ -79,6 +79,17 @@ public interface InlineIndexKeyType {
      */
     @Nullable public IndexKey get(long pageAddr, int off, int maxSize);
 
+    /**
+     * Check if inlined index key is null.
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexKeyType.java:83]: Check -> Checks

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexKeyType.java [hunk 7]
```diff
@@ -79,6 +79,17 @@ public interface InlineIndexKeyType {
...
+     *
+     * @param pageAddr Page address.
+     * @param off Offset.
+     * @param maxSize Max size.
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexKeyType.java:87]: Maybe 'maxSize' mighe be explained.
